warning: in the working copy of 'src\third_party\minimp3\minimp3.h', CRLF will be replaced by LF the next time Git touches it
diff --git "a/minimp3.h" "b/minimp3.h"
index 9fbd4d2cc..e0bc32c38 100644
--- "a/minimp3.h"
+++ "b/minimp3.h"
@@ -1,50 +1,75 @@
 #ifndef MINIMP3_H
 #define MINIMP3_H
+// SPDX-License-Identifier: CC0-1.0
 /*
     https://github.com/lieff/minimp3
     To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the public domain worldwide.
     This software is distributed without any warranty.
     See <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
-#include <stdint.h>
+#include "fl/stl/cstring.h"
+#include "fl/stl/noexcept.h"
+#include "fl/stl/stdint.h"

 #define MINIMP3_MAX_SAMPLES_PER_FRAME (1152*2)
+#define MINIMP3_SCRATCH_SIZE 16384
+
+#ifdef __cplusplus
+namespace fl {
+namespace third_party {
+#endif

 typedef struct
 {
     int frame_bytes, frame_offset, channels, hz, layer, bitrate_kbps;
 } mp3dec_frame_info_t;

-typedef struct
+typedef struct mp3dec_t
 {
     float mdct_overlap[2][9*32], qmf_state[15*2*32];
     int reserv, free_format_bytes;
     unsigned char header[4], reserv_buf[511];
 } mp3dec_t;

+typedef union mp3dec_scratch_t
+{
+    uint64_t alignment;
+    uint8_t buffer[MINIMP3_SCRATCH_SIZE];
+} mp3dec_scratch_t;
+
 #ifdef __cplusplus
 extern "C" {
 #endif /* __cplusplus */

-void mp3dec_init(mp3dec_t *dec);
+void mp3dec_init(mp3dec_t *dec) FL_NO_EXCEPT;
 #ifndef MINIMP3_FLOAT_OUTPUT
 typedef int16_t mp3d_sample_t;
 #else /* MINIMP3_FLOAT_OUTPUT */
 typedef float mp3d_sample_t;
-void mp3dec_f32_to_s16(const float *in, int16_t *out, int num_samples);
+void mp3dec_f32_to_s16(const float *in, int16_t *out, int num_samples) FL_NO_EXCEPT;
 #endif /* MINIMP3_FLOAT_OUTPUT */
-int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes, mp3d_sample_t *pcm, mp3dec_frame_info_t *info);
+int mp3dec_decode_frame_r(mp3dec_t *dec, mp3dec_scratch_t *scratch,
+                          const uint8_t *mp3, int mp3_bytes,
+                          mp3d_sample_t *pcm,
+                          mp3dec_frame_info_t *info) FL_NO_EXCEPT;
+int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes,
+                        mp3d_sample_t *pcm,
+                        mp3dec_frame_info_t *info) FL_NO_EXCEPT;

 #ifdef __cplusplus
 }
+} /* namespace third_party */
+} /* namespace fl */
 #endif /* __cplusplus */

 #endif /* MINIMP3_H */
 #if defined(MINIMP3_IMPLEMENTATION) && !defined(_MINIMP3_IMPLEMENTATION_GUARD)
 #define _MINIMP3_IMPLEMENTATION_GUARD

-#include <stdlib.h>
-#include <string.h>
+#ifdef __cplusplus
+namespace fl {
+namespace third_party {
+#endif

 #define MAX_FREE_FORMAT_FRAME_SIZE  2304    /* more than ISO spec's */
 #ifndef MAX_FRAME_SYNC_MATCHES
@@ -136,7 +161,7 @@ static __inline__ __attribute__((always_inline)) void minimp3_cpuid(int CPUInfo[
 #endif /* defined(__PIC__)*/
 }
 #endif /* defined(_MSC_VER) || defined(MINIMP3_ONLY_SIMD) */
-static int have_simd(void)
+static int have_simd(void) FL_NO_EXCEPT
 {
 #ifdef MINIMP3_ONLY_SIMD
     return 1;
@@ -236,16 +261,23 @@ typedef struct
     L3_gr_info_t gr_info[4];
     float grbuf[2][576], scf[40], syn[18 + 15][2*32];
     uint8_t ist_pos[2][39];
-} mp3dec_scratch_t;
+} mp3dec_scratch_internal_t;

-static void bs_init(bs_t *bs, const uint8_t *data, int bytes)
+#ifdef __cplusplus
+static_assert(sizeof(mp3dec_scratch_internal_t) <= MINIMP3_SCRATCH_SIZE,
+              "MINIMP3_SCRATCH_SIZE is too small");
+static_assert(alignof(mp3dec_scratch_internal_t) <= alignof(mp3dec_scratch_t),
+              "mp3dec_scratch_t alignment is too small");
+#endif
+
+static void bs_init(bs_t *bs, const uint8_t *data, int bytes) FL_NO_EXCEPT
 {
     bs->buf   = data;
     bs->pos   = 0;
     bs->limit = bytes*8;
 }

-static uint32_t get_bits(bs_t *bs, int n)
+static uint32_t get_bits(bs_t *bs, int n) FL_NO_EXCEPT
 {
     uint32_t next, cache = 0, s = bs->pos & 7;
     int shl = n + s;
@@ -261,7 +293,7 @@ static uint32_t get_bits(bs_t *bs, int n)
     return cache | (next >> -shl);
 }

-static int hdr_valid(const uint8_t *h)
+static int hdr_valid(const uint8_t *h) FL_NO_EXCEPT
 {
     return h[0] == 0xff &&
         ((h[1] & 0xF0) == 0xf0 || (h[1] & 0xFE) == 0xe2) &&
@@ -270,7 +302,7 @@ static int hdr_valid(const uint8_t *h)
         (HDR_GET_SAMPLE_RATE(h) != 3);
 }

-static int hdr_compare(const uint8_t *h1, const uint8_t *h2)
+static int hdr_compare(const uint8_t *h1, const uint8_t *h2) FL_NO_EXCEPT
 {
     return hdr_valid(h2) &&
         ((h1[1] ^ h2[1]) & 0xFE) == 0 &&
@@ -278,7 +310,7 @@ static int hdr_compare(const uint8_t *h1, const uint8_t *h2)
         !(HDR_IS_FREE_FORMAT(h1) ^ HDR_IS_FREE_FORMAT(h2));
 }

-static unsigned hdr_bitrate_kbps(const uint8_t *h)
+static unsigned hdr_bitrate_kbps(const uint8_t *h) FL_NO_EXCEPT
 {
     static const uint8_t halfrate[2][3][15] = {
         { { 0,4,8,12,16,20,24,28,32,40,48,56,64,72,80 }, { 0,4,8,12,16,20,24,28,32,40,48,56,64,72,80 }, { 0,16,24,28,32,40,48,56,64,72,80,88,96,112,128 } },
@@ -287,18 +319,18 @@ static unsigned hdr_bitrate_kbps(const uint8_t *h)
     return 2*halfrate[!!HDR_TEST_MPEG1(h)][HDR_GET_LAYER(h) - 1][HDR_GET_BITRATE(h)];
 }

-static unsigned hdr_sample_rate_hz(const uint8_t *h)
+static unsigned hdr_sample_rate_hz(const uint8_t *h) FL_NO_EXCEPT
 {
     static const unsigned g_hz[3] = { 44100, 48000, 32000 };
     return g_hz[HDR_GET_SAMPLE_RATE(h)] >> (int)!HDR_TEST_MPEG1(h) >> (int)!HDR_TEST_NOT_MPEG25(h);
 }

-static unsigned hdr_frame_samples(const uint8_t *h)
+static unsigned hdr_frame_samples(const uint8_t *h) FL_NO_EXCEPT
 {
     return HDR_IS_LAYER_1(h) ? 384 : (1152 >> (int)HDR_IS_FRAME_576(h));
 }

-static int hdr_frame_bytes(const uint8_t *h, int free_format_size)
+static int hdr_frame_bytes(const uint8_t *h, int free_format_size) FL_NO_EXCEPT
 {
     int frame_bytes = hdr_frame_samples(h)*hdr_bitrate_kbps(h)*125/hdr_sample_rate_hz(h);
     if (HDR_IS_LAYER_1(h))
@@ -308,13 +340,13 @@ static int hdr_frame_bytes(const uint8_t *h, int free_format_size)
     return frame_bytes ? frame_bytes : free_format_size;
 }

-static int hdr_padding(const uint8_t *h)
+static int hdr_padding(const uint8_t *h) FL_NO_EXCEPT
 {
     return HDR_TEST_PADDING(h) ? (HDR_IS_LAYER_1(h) ? 4 : 1) : 0;
 }

 #ifndef MINIMP3_ONLY_MP3
-static const L12_subband_alloc_t *L12_subband_alloc_table(const uint8_t *hdr, L12_scale_info *sci)
+static const L12_subband_alloc_t *L12_subband_alloc_table(const uint8_t *hdr, L12_scale_info *sci) FL_NO_EXCEPT
 {
     const L12_subband_alloc_t *alloc;
     int mode = HDR_GET_STEREO_MODE(hdr);
@@ -359,7 +391,7 @@ static const L12_subband_alloc_t *L12_subband_alloc_table(const uint8_t *hdr, L1
     return alloc;
 }

-static void L12_read_scalefactors(bs_t *bs, uint8_t *pba, uint8_t *scfcod, int bands, float *scf)
+static void L12_read_scalefactors(bs_t *bs, uint8_t *pba, uint8_t *scfcod, int bands, float *scf) FL_NO_EXCEPT
 {
     static const float g_deq_L12[18*3] = {
 #define DQ(x) 9.53674316e-07f/x, 7.56931807e-07f/x, 6.00777173e-07f/x
@@ -383,7 +415,7 @@ static void L12_read_scalefactors(bs_t *bs, uint8_t *pba, uint8_t *scfcod, int b
     }
 }

-static void L12_read_scale_info(const uint8_t *hdr, bs_t *bs, L12_scale_info *sci)
+static void L12_read_scale_info(const uint8_t *hdr, bs_t *bs, L12_scale_info *sci) FL_NO_EXCEPT
 {
     static const uint8_t g_bitalloc_code_tab[] = {
         0,17, 3, 4, 5,6,7, 8,9,10,11,12,13,14,15,16,
@@ -431,7 +463,7 @@ static void L12_read_scale_info(const uint8_t *hdr, bs_t *bs, L12_scale_info *sc
     }
 }

-static int L12_dequantize_granule(float *grbuf, bs_t *bs, L12_scale_info *sci, int group_size)
+static int L12_dequantize_granule(float *grbuf, bs_t *bs, L12_scale_info *sci, int group_size) FL_NO_EXCEPT
 {
     int i, j, k, choff = 576;
     for (j = 0; j < 4; j++)
@@ -466,7 +498,7 @@ static int L12_dequantize_granule(float *grbuf, bs_t *bs, L12_scale_info *sci, i
     return group_size*4;
 }

-static void L12_apply_scf_384(L12_scale_info *sci, const float *scf, float *dst)
+static void L12_apply_scf_384(L12_scale_info *sci, const float *scf, float *dst) FL_NO_EXCEPT
 {
     int i, k;
     memcpy(dst + 576 + sci->stereo_bands*18, dst + sci->stereo_bands*18, (sci->total_bands - sci->stereo_bands)*18*sizeof(float));
@@ -481,7 +513,7 @@ static void L12_apply_scf_384(L12_scale_info *sci, const float *scf, float *dst)
 }
 #endif /* MINIMP3_ONLY_MP3 */

-static int L3_read_side_info(bs_t *bs, L3_gr_info_t *gr, const uint8_t *hdr)
+static int L3_read_side_info(bs_t *bs, L3_gr_info_t *gr, const uint8_t *hdr) FL_NO_EXCEPT
 {
     static const uint8_t g_scf_long[8][23] = {
         { 6,6,6,6,6,6,8,10,12,14,16,20,24,28,32,38,46,52,60,68,58,54,0 },
@@ -606,7 +638,7 @@ static int L3_read_side_info(bs_t *bs, L3_gr_info_t *gr, const uint8_t *hdr)
     return main_data_begin;
 }

-static void L3_read_scalefactors(uint8_t *scf, uint8_t *ist_pos, const uint8_t *scf_size, const uint8_t *scf_count, bs_t *bitbuf, int scfsi)
+static void L3_read_scalefactors(uint8_t *scf, uint8_t *ist_pos, const uint8_t *scf_size, const uint8_t *scf_count, bs_t *bitbuf, int scfsi) FL_NO_EXCEPT
 {
     int i, k;
     for (i = 0; i < 4 && scf_count[i]; i++, scfsi *= 2)
@@ -639,7 +671,7 @@ static void L3_read_scalefactors(uint8_t *scf, uint8_t *ist_pos, const uint8_t *
     scf[0] = scf[1] = scf[2] = 0;
 }

-static float L3_ldexp_q2(float y, int exp_q2)
+static float L3_ldexp_q2(float y, int exp_q2) FL_NO_EXCEPT
 {
     static const float g_expfrac[4] = { 9.31322575e-10f,7.83145814e-10f,6.58544508e-10f,5.53767716e-10f };
     int e;
@@ -651,7 +683,7 @@ static float L3_ldexp_q2(float y, int exp_q2)
     return y;
 }

-static void L3_decode_scalefactors(const uint8_t *hdr, uint8_t *ist_pos, bs_t *bs, const L3_gr_info_t *gr, float *scf, int ch)
+static void L3_decode_scalefactors(const uint8_t *hdr, uint8_t *ist_pos, bs_t *bs, const L3_gr_info_t *gr, float *scf, int ch) FL_NO_EXCEPT
 {
     static const uint8_t g_scf_partitions[3][28] = {
         { 6,5,5, 5,6,5,5,5,6,5, 7,3,11,10,0,0, 7, 7, 7,0, 6, 6,6,3, 8, 8,5,0 },
@@ -718,7 +750,7 @@ static const float g_pow43[129 + 16] = {
     0,1,2.519842f,4.326749f,6.349604f,8.549880f,10.902724f,13.390518f,16.000000f,18.720754f,21.544347f,24.463781f,27.473142f,30.567351f,33.741992f,36.993181f,40.317474f,43.711787f,47.173345f,50.699631f,54.288352f,57.937408f,61.644865f,65.408941f,69.227979f,73.100443f,77.024898f,81.000000f,85.024491f,89.097188f,93.216975f,97.382800f,101.593667f,105.848633f,110.146801f,114.487321f,118.869381f,123.292209f,127.755065f,132.257246f,136.798076f,141.376907f,145.993119f,150.646117f,155.335327f,160.060199f,164.820202f,169.614826f,174.443577f,179.305980f,184.201575f,189.129918f,194.090580f,199.083145f,204.107210f,209.162385f,214.248292f,219.364564f,224.510845f,229.686789f,234.892058f,240.126328f,245.389280f,250.680604f,256.000000f,261.347174f,266.721841f,272.123723f,277.552547f,283.008049f,288.489971f,293.998060f,299.532071f,305.091761f,310.676898f,316.287249f,321.922592f,327.582707f,333.267377f,338.976394f,344.709550f,350.466646f,356.247482f,362.051866f,367.879608f,373.730522f,379.604427f,385.501143f,391.420496f,397.362314f,403.326427f,409.312672f,415.320884f,421.350905f,427.402579f,433.475750f,439.570269f,445.685987f,451.822757f,457.980436f,464.158883f,470.357960f,476.577530f,482.817459f,489.077615f,495.357868f,501.658090f,507.978156f,514.317941f,520.677324f,527.056184f,533.454404f,539.871867f,546.308458f,552.764065f,559.238575f,565.731879f,572.243870f,578.774440f,585.323483f,591.890898f,598.476581f,605.080431f,611.702349f,618.342238f,625.000000f,631.675540f,638.368763f,645.079578f
 };

-static float L3_pow_43(int x)
+static float L3_pow_43(int x) FL_NO_EXCEPT
 {
     float frac;
     int sign, mult = 256;
@@ -739,7 +771,7 @@ static float L3_pow_43(int x)
     return g_pow43[16 + ((x + sign) >> 6)]*(1.f + frac*((4.f/3) + frac*(2.f/9)))*mult;
 }

-static void L3_huffman(float *dst, bs_t *bs, const L3_gr_info_t *gr_info, const float *scf, int layer3gr_limit)
+static void L3_huffman(float *dst, bs_t *bs, const L3_gr_info_t *gr_info, const float *scf, int layer3gr_limit) FL_NO_EXCEPT
 {
     static const int16_t tabs[] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         785,785,785,785,784,784,784,784,513,513,513,513,513,513,513,513,256,256,256,256,256,256,256,256,256,256,256,256,256,256,256,256,
@@ -876,7 +908,7 @@ static void L3_huffman(float *dst, bs_t *bs, const L3_gr_info_t *gr_info, const
     bs->pos = layer3gr_limit;
 }

-static void L3_midside_stereo(float *left, int n)
+static void L3_midside_stereo(float *left, int n) FL_NO_EXCEPT
 {
     int i = 0;
     float *right = left + 576;
@@ -908,7 +940,7 @@ static void L3_midside_stereo(float *left, int n)
     }
 }

-static void L3_intensity_stereo_band(float *left, int n, float kl, float kr)
+static void L3_intensity_stereo_band(float *left, int n, float kl, float kr) FL_NO_EXCEPT
 {
     int i;
     for (i = 0; i < n; i++)
@@ -918,7 +950,7 @@ static void L3_intensity_stereo_band(float *left, int n, float kl, float kr)
     }
 }

-static void L3_stereo_top_band(const float *right, const uint8_t *sfb, int nbands, int max_band[3])
+static void L3_stereo_top_band(const float *right, const uint8_t *sfb, int nbands, int max_band[3]) FL_NO_EXCEPT // ok array parameter
 {
     int i, k;

@@ -938,7 +970,7 @@ static void L3_stereo_top_band(const float *right, const uint8_t *sfb, int nband
     }
 }

-static void L3_stereo_process(float *left, const uint8_t *ist_pos, const uint8_t *sfb, const uint8_t *hdr, int max_band[3], int mpeg2_sh)
+static void L3_stereo_process(float *left, const uint8_t *ist_pos, const uint8_t *sfb, const uint8_t *hdr, int max_band[3], int mpeg2_sh) FL_NO_EXCEPT // ok array parameter
 {
     static const float g_pan[7*2] = { 0,1,0.21132487f,0.78867513f,0.36602540f,0.63397460f,0.5f,0.5f,0.63397460f,0.36602540f,0.78867513f,0.21132487f,1,0 };
     unsigned i, max_pos = HDR_TEST_MPEG1(hdr) ? 7 : 64;
@@ -972,7 +1004,7 @@ static void L3_stereo_process(float *left, const uint8_t *ist_pos, const uint8_t
     }
 }

-static void L3_intensity_stereo(float *left, uint8_t *ist_pos, const L3_gr_info_t *gr, const uint8_t *hdr)
+static void L3_intensity_stereo(float *left, uint8_t *ist_pos, const L3_gr_info_t *gr, const uint8_t *hdr) FL_NO_EXCEPT
 {
     int max_band[3], n_sfb = gr->n_long_sfb + gr->n_short_sfb;
     int i, max_blocks = gr->n_short_sfb ? 3 : 1;
@@ -992,7 +1024,7 @@ static void L3_intensity_stereo(float *left, uint8_t *ist_pos, const L3_gr_info_
     L3_stereo_process(left, ist_pos, gr->sfbtab, hdr, max_band, gr[1].scalefac_compress & 1);
 }

-static void L3_reorder(float *grbuf, float *scratch, const uint8_t *sfb)
+static void L3_reorder(float *grbuf, float *scratch, const uint8_t *sfb) FL_NO_EXCEPT
 {
     int i, len;
     float *src = grbuf, *dst = scratch;
@@ -1009,7 +1041,7 @@ static void L3_reorder(float *grbuf, float *scratch, const uint8_t *sfb)
     memcpy(grbuf, scratch, (dst - scratch)*sizeof(float));
 }

-static void L3_antialias(float *grbuf, int nbands)
+static void L3_antialias(float *grbuf, int nbands) FL_NO_EXCEPT
 {
     static const float g_aa[2][8] = {
         {0.85749293f,0.88174200f,0.94962865f,0.98331459f,0.99551782f,0.99916056f,0.99989920f,0.99999316f},
@@ -1044,7 +1076,7 @@ static void L3_antialias(float *grbuf, int nbands)
     }
 }

-static void L3_dct3_9(float *y)
+static void L3_dct3_9(float *y) FL_NO_EXCEPT
 {
     float s0, s1, s2, s3, s4, s5, s6, s7, s8, t0, t2, t4;

@@ -1084,7 +1116,7 @@ static void L3_dct3_9(float *y)
     y[8] = s4 + s7;
 }

-static void L3_imdct36(float *grbuf, float *overlap, const float *window, int nbands)
+static void L3_imdct36(float *grbuf, float *overlap, const float *window, int nbands) FL_NO_EXCEPT
 {
     int i, j;
     static const float g_twid9[18] = {
@@ -1141,7 +1173,7 @@ static void L3_imdct36(float *grbuf, float *overlap, const float *window, int nb
     }
 }

-static void L3_idct3(float x0, float x1, float x2, float *dst)
+static void L3_idct3(float x0, float x1, float x2, float *dst) FL_NO_EXCEPT
 {
     float m1 = x1*0.86602540f;
     float a1 = x0 - x2*0.5f;
@@ -1150,7 +1182,7 @@ static void L3_idct3(float x0, float x1, float x2, float *dst)
     dst[2] = a1 - m1;
 }

-static void L3_imdct12(float *x, float *dst, float *overlap)
+static void L3_imdct12(float *x, float *dst, float *overlap) FL_NO_EXCEPT
 {
     static const float g_twid3[6] = { 0.79335334f,0.92387953f,0.99144486f, 0.60876143f,0.38268343f,0.13052619f };
     float co[3], si[3];
@@ -1170,7 +1202,7 @@ static void L3_imdct12(float *x, float *dst, float *overlap)
     }
 }

-static void L3_imdct_short(float *grbuf, float *overlap, int nbands)
+static void L3_imdct_short(float *grbuf, float *overlap, int nbands) FL_NO_EXCEPT
 {
     for (;nbands > 0; nbands--, overlap += 9, grbuf += 18)
     {
@@ -1183,7 +1215,7 @@ static void L3_imdct_short(float *grbuf, float *overlap, int nbands)
     }
 }

-static void L3_change_sign(float *grbuf)
+static void L3_change_sign(float *grbuf) FL_NO_EXCEPT
 {
     int b, i;
     for (b = 0, grbuf += 18; b < 32; b += 2, grbuf += 36)
@@ -1191,7 +1223,7 @@ static void L3_change_sign(float *grbuf)
             grbuf[i] = -grbuf[i];
 }

-static void L3_imdct_gr(float *grbuf, float *overlap, unsigned block_type, unsigned n_long_bands)
+static void L3_imdct_gr(float *grbuf, float *overlap, unsigned block_type, unsigned n_long_bands) FL_NO_EXCEPT
 {
     static const float g_mdct_window[2][18] = {
         { 0.99904822f,0.99144486f,0.97629601f,0.95371695f,0.92387953f,0.88701083f,0.84339145f,0.79335334f,0.73727734f,0.04361938f,0.13052619f,0.21643961f,0.30070580f,0.38268343f,0.46174861f,0.53729961f,0.60876143f,0.67559021f },
@@ -1209,7 +1241,7 @@ static void L3_imdct_gr(float *grbuf, float *overlap, unsigned block_type, unsig
         L3_imdct36(grbuf, overlap, g_mdct_window[block_type == STOP_BLOCK_TYPE], 32 - n_long_bands);
 }

-static void L3_save_reservoir(mp3dec_t *h, mp3dec_scratch_t *s)
+static void L3_save_reservoir(mp3dec_t *h, mp3dec_scratch_internal_t *s) FL_NO_EXCEPT
 {
     int pos = (s->bs.pos + 7)/8u;
     int remains = s->bs.limit/8u - pos;
@@ -1225,7 +1257,7 @@ static void L3_save_reservoir(mp3dec_t *h, mp3dec_scratch_t *s)
     h->reserv = remains;
 }

-static int L3_restore_reservoir(mp3dec_t *h, bs_t *bs, mp3dec_scratch_t *s, int main_data_begin)
+static int L3_restore_reservoir(mp3dec_t *h, bs_t *bs, mp3dec_scratch_internal_t *s, int main_data_begin) FL_NO_EXCEPT
 {
     int frame_bytes = (bs->limit - bs->pos)/8;
     int bytes_have = MINIMP3_MIN(h->reserv, main_data_begin);
@@ -1235,7 +1267,7 @@ static int L3_restore_reservoir(mp3dec_t *h, bs_t *bs, mp3dec_scratch_t *s, int
     return h->reserv >= main_data_begin;
 }

-static void L3_decode(mp3dec_t *h, mp3dec_scratch_t *s, L3_gr_info_t *gr_info, int nch)
+static void L3_decode(mp3dec_t *h, mp3dec_scratch_internal_t *s, L3_gr_info_t *gr_info, int nch) FL_NO_EXCEPT
 {
     int ch;

@@ -1271,7 +1303,7 @@ static void L3_decode(mp3dec_t *h, mp3dec_scratch_t *s, L3_gr_info_t *gr_info, i
     }
 }

-static void mp3d_DCT_II(float *grbuf, int n)
+static void mp3d_DCT_II(float *grbuf, int n) FL_NO_EXCEPT
 {
     static const float g_sec[24] = {
         10.19000816f,0.50060302f,0.50241929f,3.40760851f,0.50547093f,0.52249861f,2.05778098f,0.51544732f,0.56694406f,1.48416460f,0.53104258f,0.64682180f,1.16943991f,0.55310392f,0.78815460f,0.97256821f,0.58293498f,1.06067765f,0.83934963f,0.62250412f,1.72244716f,0.74453628f,0.67480832f,5.10114861f
@@ -1427,7 +1459,7 @@ static void mp3d_DCT_II(float *grbuf, int n)
 }

 #ifndef MINIMP3_FLOAT_OUTPUT
-static int16_t mp3d_scale_pcm(float sample)
+static int16_t mp3d_scale_pcm(float sample) FL_NO_EXCEPT
 {
 #if HAVE_ARMV6
     int32_t s32 = (int32_t)(sample + .5f);
@@ -1448,7 +1480,7 @@ static float mp3d_scale_pcm(float sample)
 }
 #endif /* MINIMP3_FLOAT_OUTPUT */

-static void mp3d_synth_pair(mp3d_sample_t *pcm, int nch, const float *z)
+static void mp3d_synth_pair(mp3d_sample_t *pcm, int nch, const float *z) FL_NO_EXCEPT
 {
     float a;
     a  = (z[14*64] - z[    0]) * 29;
@@ -1473,7 +1505,7 @@ static void mp3d_synth_pair(mp3d_sample_t *pcm, int nch, const float *z)
     pcm[16*nch] = mp3d_scale_pcm(a);
 }

-static void mp3d_synth(float *xl, mp3d_sample_t *dstl, int nch, float *lins)
+static void mp3d_synth(float *xl, mp3d_sample_t *dstl, int nch, float *lins) FL_NO_EXCEPT
 {
     int i;
     float *xr = xl + 576*(nch - 1);
@@ -1626,7 +1658,7 @@ static void mp3d_synth(float *xl, mp3d_sample_t *dstl, int nch, float *lins)
 #endif /* MINIMP3_ONLY_SIMD */
 }

-static void mp3d_synth_granule(float *qmf_state, float *grbuf, int nbands, int nch, mp3d_sample_t *pcm, float *lins)
+static void mp3d_synth_granule(float *qmf_state, float *grbuf, int nbands, int nch, mp3d_sample_t *pcm, float *lins) FL_NO_EXCEPT
 {
     int i;
     for (i = 0; i < nch; i++)
@@ -1654,7 +1686,7 @@ static void mp3d_synth_granule(float *qmf_state, float *grbuf, int nbands, int n
     }
 }

-static int mp3d_match_frame(const uint8_t *hdr, int mp3_bytes, int frame_bytes)
+static int mp3d_match_frame(const uint8_t *hdr, int mp3_bytes, int frame_bytes) FL_NO_EXCEPT
 {
     int i, nmatch;
     for (i = 0, nmatch = 0; nmatch < MAX_FRAME_SYNC_MATCHES; nmatch++)
@@ -1668,7 +1700,7 @@ static int mp3d_match_frame(const uint8_t *hdr, int mp3_bytes, int frame_bytes)
     return 1;
 }

-static int mp3d_find_frame(const uint8_t *mp3, int mp3_bytes, int *free_format_bytes, int *ptr_frame_bytes)
+static int mp3d_find_frame(const uint8_t *mp3, int mp3_bytes, int *free_format_bytes, int *ptr_frame_bytes) FL_NO_EXCEPT
 {
     int i, k;
     for (i = 0; i < mp3_bytes - HDR_SIZE; i++, mp3++)
@@ -1705,17 +1737,21 @@ static int mp3d_find_frame(const uint8_t *mp3, int mp3_bytes, int *free_format_b
     return mp3_bytes;
 }

-void mp3dec_init(mp3dec_t *dec)
+void mp3dec_init(mp3dec_t *dec) FL_NO_EXCEPT
 {
     dec->header[0] = 0;
 }

-int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes, mp3d_sample_t *pcm, mp3dec_frame_info_t *info)
+int mp3dec_decode_frame_r(mp3dec_t *dec, mp3dec_scratch_t *scratch_storage,
+                          const uint8_t *mp3, int mp3_bytes,
+                          mp3d_sample_t *pcm,
+                          mp3dec_frame_info_t *info) FL_NO_EXCEPT
 {
     int i = 0, igr, frame_size = 0, success = 1;
     const uint8_t *hdr;
     bs_t bs_frame[1];
-    mp3dec_scratch_t scratch;
+    mp3dec_scratch_internal_t *scratch =
+        (mp3dec_scratch_internal_t *)(void *)scratch_storage->buffer;

     if (mp3_bytes > 4 && dec->header[0] == 0xff && hdr_compare(dec->header, mp3))
     {
@@ -1758,23 +1794,23 @@ int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes, mp3d_s

     if (info->layer == 3)
     {
-        int main_data_begin = L3_read_side_info(bs_frame, scratch.gr_info, hdr);
+        int main_data_begin = L3_read_side_info(bs_frame, scratch->gr_info, hdr);
         if (main_data_begin < 0 || bs_frame->pos > bs_frame->limit)
         {
             mp3dec_init(dec);
             return 0;
         }
-        success = L3_restore_reservoir(dec, bs_frame, &scratch, main_data_begin);
+        success = L3_restore_reservoir(dec, bs_frame, scratch, main_data_begin);
         if (success)
         {
             for (igr = 0; igr < (HDR_TEST_MPEG1(hdr) ? 2 : 1); igr++, pcm += 576*info->channels)
             {
-                memset(scratch.grbuf[0], 0, 576*2*sizeof(float));
-                L3_decode(dec, &scratch, scratch.gr_info + igr*info->channels, info->channels);
-                mp3d_synth_granule(dec->qmf_state, scratch.grbuf[0], 18, info->channels, pcm, scratch.syn[0]);
+                memset(scratch->grbuf[0], 0, 576*2*sizeof(float));
+                L3_decode(dec, scratch, scratch->gr_info + igr*info->channels, info->channels);
+                mp3d_synth_granule(dec->qmf_state, scratch->grbuf[0], 18, info->channels, pcm, scratch->syn[0]);
             }
         }
-        L3_save_reservoir(dec, &scratch);
+        L3_save_reservoir(dec, scratch);
     } else
     {
 #ifdef MINIMP3_ONLY_MP3
@@ -1783,15 +1819,15 @@ int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes, mp3d_s
         L12_scale_info sci[1];
         L12_read_scale_info(hdr, bs_frame, sci);

-        memset(scratch.grbuf[0], 0, 576*2*sizeof(float));
+        memset(scratch->grbuf[0], 0, 576*2*sizeof(float));
         for (i = 0, igr = 0; igr < 3; igr++)
         {
-            if (12 == (i += L12_dequantize_granule(scratch.grbuf[0] + i, bs_frame, sci, info->layer | 1)))
+            if (12 == (i += L12_dequantize_granule(scratch->grbuf[0] + i, bs_frame, sci, info->layer | 1)))
             {
                 i = 0;
-                L12_apply_scf_384(sci, sci->scf + igr, scratch.grbuf[0]);
-                mp3d_synth_granule(dec->qmf_state, scratch.grbuf[0], 12, info->channels, pcm, scratch.syn[0]);
-                memset(scratch.grbuf[0], 0, 576*2*sizeof(float));
+                L12_apply_scf_384(sci, sci->scf + igr, scratch->grbuf[0]);
+                mp3d_synth_granule(dec->qmf_state, scratch->grbuf[0], 12, info->channels, pcm, scratch->syn[0]);
+                memset(scratch->grbuf[0], 0, 576*2*sizeof(float));
                 pcm += 384*info->channels;
             }
             if (bs_frame->pos > bs_frame->limit)
@@ -1805,8 +1841,16 @@ int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes, mp3d_s
     return success*hdr_frame_samples(dec->header);
 }

+int mp3dec_decode_frame(mp3dec_t *dec, const uint8_t *mp3, int mp3_bytes,
+                        mp3d_sample_t *pcm,
+                        mp3dec_frame_info_t *info) FL_NO_EXCEPT
+{
+    mp3dec_scratch_t scratch;
+    return mp3dec_decode_frame_r(dec, &scratch, mp3, mp3_bytes, pcm, info);
+}
+
 #ifdef MINIMP3_FLOAT_OUTPUT
-void mp3dec_f32_to_s16(const float *in, int16_t *out, int num_samples)
+void mp3dec_f32_to_s16(const float *in, int16_t *out, int num_samples) FL_NO_EXCEPT
 {
     int i = 0;
 #if HAVE_SIMD
@@ -1862,4 +1906,8 @@ void mp3dec_f32_to_s16(const float *in, int16_t *out, int num_samples)
     }
 }
 #endif /* MINIMP3_FLOAT_OUTPUT */
+#ifdef __cplusplus
+} /* namespace third_party */
+} /* namespace fl */
+#endif
 #endif /* MINIMP3_IMPLEMENTATION && !_MINIMP3_IMPLEMENTATION_GUARD */
